### PR TITLE
[MSHARED-1010] Allow to set dedicated Maven home directory

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

**Comment by kwin** [comment]
@michael-o or @rfscholte : Can you have a look?

File: .gitignore [hunk 1]
```diff
@@ -3,7 +3,7 @@ target/
 .classpath
 .settings/
 .svn/
-bin/
+/bin/
```
Comment (michael-o) [inline .gitignore:6]: ???

**Review by michael-o — COMMENTED** [review]
This change confuses me, it seems that it contains unrelated changes. If so, move them to a separate preceeding PR.

File: .gitignore (hunk 1 above)
Comment (kwin) [reply .gitignore:6]: a bin folder is included in the test resources as this is part of the layout below maven home! The intent here is to only exclude compiled classes which will only ever end up in the to-level directory /bin (if at all)!

**Comment by kwin** [comment]
> it seems that it contains unrelated changes. 

which changes are you referring to exactly?

File: src/main/java/org/apache/maven/it/Verifier.java [hunk 2]
```diff
@@ -103,23 +103,32 @@
 
     private boolean debug;
 
+    /** 
+     * If {@code true} uses {@link ForkedLauncher}, if {@code false} uses {@link Embedded3xLauncher},
+     * otherwise considers the value {@link #forkMode}.
+     */
```
Comment (michael-o) [inline src/main/java/org/apache/maven/it/Verifier.java:109]: This

File: src/main/java/org/apache/maven/it/Verifier.java [hunk 3]
```diff
@@ -103,23 +103,32 @@
...
     // will launch mvn with --debug 
     private boolean mavenDebug = false;
 
+    /**
+     * Either "auto" (use {@link ForkedLauncher} when {@link #environmentVariables} is not empty,
+     * otherwise use {@link Embedded3xLauncher}) , "embedder" (always use {@link Embedded3xLauncher})
+     * or something else (always use {@link ForkedLauncher}).
+     * Set through system property {@code verifier.forkMode}.
+     * Only relevant if {@link #forkJvm} is {@code null}.
+     */
```
Comment (michael-o) [inline src/main/java/org/apache/maven/it/Verifier.java:125]: This

File: src/main/java/org/apache/maven/it/Verifier.java [hunk 4]
```diff
@@ -103,23 +103,32 @@
...
     private String forkMode;
 
     private boolean debugJvm = false;
 
+    private boolean useWrapper;
```
Comment (michael-o) [inline src/main/java/org/apache/maven/it/Verifier.java:130]: This

File: src/main/java/org/apache/maven/it/Verifier.java [hunk 5]
```diff
@@ -182,22 +203,28 @@ private Verifier( String basedir, String settingsFile, boolean debug, Boolean fo
         setDebug( debug );
 
         findLocalRepo( settingsFile );
-        findDefaultMavenHome();
+        if ( mavenHome == null )
+        {
+            this.mavenHome = getDefaultMavenHome();
+            useWrapper = Files.exists( Paths.get( getBasedir(), "mvnw" ) );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/it/Verifier.java:209]: This

File: src/main/java/org/apache/maven/it/Verifier.java (hunk 2 above)
Comment (kwin) [reply src/main/java/org/apache/maven/it/Verifier.java:109]: Sorry but I don't think that adding javadoc to existing/unmodified fields/methods deserves a dedicated JIRA nor PR. This was mostly for me to understand how the functionality works. I can remove it if you want, but I won't dedicated time to recontribute that again in another PR.

File: src/main/java/org/apache/maven/it/Verifier.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/maven/it/Verifier.java:125]: Same as above

File: src/main/java/org/apache/maven/it/Verifier.java (hunk 4 above)
Comment (kwin) [reply src/main/java/org/apache/maven/it/Verifier.java:130]: Obviously when using a dedicated maven home no `mvnw` should be used, therefore this required to determine this in the constructor and introducing this field.

File: src/main/java/org/apache/maven/it/Verifier.java (hunk 5 above)
Comment (kwin) [reply src/main/java/org/apache/maven/it/Verifier.java:209]: See above

**Comment by michael-o** [comment]
> > it seems that it contains unrelated changes.
> 
> which changes are you referring to exactly?

I am fine with your explanations.

**Review by slawekjaranowski — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #149

**Comment by jira-importer** [comment]
Resolve #149

**Comment by jira-importer** [comment]
Resolve #149

**Pull request merged**